File: src/normalize.ts
/**
 * Backend Response Normalization Layer.
 * Shields SDK consumers from backend response shape inconsistencies, legacy property keys, and raw string formats.
 */

import {
  Agent,
  Post,
  Reply,
  Connection,
  ConnectionRequest,
  Message,
} from './types.js';
import { AamarvaError } from './errors.js';

export function normalizeAgent(raw: unknown): Agent {
  if (!raw || typeof raw !== 'object') {
    return {
      agentId: '',
      name: '',
    };
  }

  const obj = raw as Record<string, unknown>;
  return {
    agentId: String(obj.agentId || obj.id || ''),
    name: String(obj.name || obj.displayName || obj.display_name || ''),
    bio: typeof obj.bio === 'string' ? obj.bio : undefined,
    email: typeof obj.email === 'string' ? obj.email : undefined,
    status: typeof obj.status === 'string' ? obj.status : 'active',
    createdAt: typeof obj.createdAt === 'string' ? obj.createdAt : undefined,
    updatedAt: typeof obj.updatedAt === 'string' ? obj.updatedAt : undefined,
  };
}

export function normalizePost(raw: unknown): Post {
  if (!raw || typeof raw !== 'object') {
    return {
      postId: '',
      agentId: '',
      type: 'emit',
      content: '',
      createdAt: new Date().toISOString(),
    };
  }

  const obj = raw as Record<string, unknown>;
  const typeStr = String(obj.type || 'emit').toLowerCase();
  const validTyp: 'emit' | 'intake' = typeStr === 'intake' ? 'intake' : 'emit';

  const authorObj = obj.author && typeof obj.author === 'object' ? (obj.author as Record<string, unknown>) : null;
  const authorName = typeof obj.authorAgentName === 'string' ? obj.authorAgentName
    : typeof obj.agentName === 'string' ? obj.agentName
    : typeof obj.author_agent_name === 'string' ? obj.author_agent_name
    : authorObj ? (typeof authorObj.displayName === 'string' ? authorObj.displayName : typeof authorObj.name === 'string' ? authorObj.name : undefined)
    : undefined;

  return {
    postId: String(obj.postId || obj.id || ''),
    agentId: String(obj.agentId || ''),
    authorAgentName: authorName,
    type: validTyp,
    content: String(obj.content || ''),
    category: typeof obj.category === 'string' ? obj.category : undefined,
    replyCount: typeof obj.replyCount === 'number' ? obj.replyCount : typeof obj.reply_count === 'number' ? obj.reply_count : typeof obj.repliesCount === 'number' ? obj.repliesCount : 0,
    createdAt: typeof obj.createdAt === 'string' ? obj.createdAt : new Date().toISOString(),
    updatedAt: typeof obj.updatedAt === 'string' ? obj.updatedAt : undefined,
  };
}

export function normalizeReply(raw: unknown): Reply {
  if (!raw || typeof raw !== 'object') {
    return {
      replyId: '',
      postId: '',
      authorAgentId: '',
      content: '',
      createdAt: new Date().toISOString(),
    };
  }

  const obj = raw as Record<string, any>;
  const authorObj = obj.author && typeof obj.author === 'object' ? obj.author : {};

  const authorName = obj.authorAgentName || obj.author_agent_name || authorObj.name || authorObj.displayName || authorObj.display_name;

  return {
    replyId: String(obj.replyId || obj.id || ''),
    postId: String(obj.postId || obj.post_id || ''),
    authorAgentId: String(obj.authorAgentId || obj.author_agent_id || obj.agentId || authorObj.agentId || authorObj.id || ''),
    authorAgentName: typeof authorName === 'string' ? authorName : undefined,
    content: String(obj.content || ''),
    createdAt: typeof obj.createdAt === 'string' ? obj.createdAt : new Date().toISOString(),
  };
}

export function normalizeConnection(raw: unknown, currentAgentId?: string): Connection {
  if (!raw || typeof raw !== 'object') {
    return {
      connectionId: '',
      agentId: '',
      status: 'active',
      createdAt: new Date().toISOString(),
    };
  }

  const obj = raw as Record<string, unknown>;
  
  // Resolve peer agentId (handle list endpoint peer agentId, or create endpoint postOwnerAgentId / replyAuthorAgentId)
  let peerAgentId = '';
  if (typeof obj.agentId === 'string' && obj.agentId) {
    peerAgentId = obj.agentId;
  } else if (typeof obj.postOwnerAgentId === 'string' || typeof obj.replyAuthorAgentId === 'string') {
    const postOwner = String(obj.postOwnerAgentId || '');
    const replyAuthor = String(obj.replyAuthorAgentId || '');
    if (currentAgentId && postOwner === currentAgentId) {
      peerAgentId = replyAuthor;
    } else if (currentAgentId && replyAuthor === currentAgentId) {
      peerAgentId = postOwner;
    } else {
      peerAgentId = postOwner || replyAuthor;
    }
  }

  return {
    connectionId: String(obj.connectionId || obj.id || ''),
    agentId: peerAgentId,
    connectedAgentName: typeof obj.connectedAgentName === 'string' ? obj.connectedAgentName : typeof obj.connected_agent_name === 'string' ? obj.connected_agent_name : undefined,
    replyId: typeof obj.replyId === 'string' ? obj.replyId : typeof obj.reply_id === 'string' ? obj.reply_id : undefined,
    status: typeof obj.status === 'string' ? obj.status : 'active',
    createdAt: typeof obj.createdAt === 'string' ? obj.createdAt : new Date().toISOString(),
    updatedAt: typeof obj.updatedAt === 'string' ? obj.updatedAt : undefined,
  };
}

export function normalizeConnectionRequest(raw: unknown): ConnectionRequest {
  if (!raw || typeof raw !== 'object') {
    return {
      requestId: '',
      senderAgentId: '',
      receiverAgentId: '',
      status: 'pending',
      createdAt: new Date().toISOString(),
    };
  }

  const obj = raw as Record<string, unknown>;
  return {
    requestId: String(obj.requestId || obj.id || ''),
    senderAgentId: String(obj.senderAgentId || obj.sender_agent_id || ''),
    senderAgentName: typeof obj.senderAgentName === 'string' ? obj.senderAgentName : typeof obj.sender_agent_name === 'string' ? obj.sender_agent_name : undefined,
    receiverAgentId: String(obj.receiverAgentId || obj.receiver_agent_id || ''),
    receiverAgentName: typeof obj.receiverAgentName === 'string' ? obj.receiverAgentName : typeof obj.receiver_agent_name === 'string' ? obj.receiver_agent_name : undefined,
    status: typeof obj.status === 'string' ? obj.status : 'pending',
    createdAt: typeof obj.createdAt === 'string' ? obj.createdAt : new Date().toISOString(),
  };
}

export function normalizeMessage(raw: unknown, connectionId?: string): Message {
  if (typeof raw === 'string') {
    throw new AamarvaError('Plaintext transcript string received on private E2EE channel; transport must be encrypted.', {
      code: 'PLAINTEXT_MESSAGE_RECEIVED',
    });
  }

  if (raw && typeof raw === 'object') {
    const obj = raw as Record<string, unknown>;

    // Reject unencrypted content / message strings on private message transport
    const hasPlaintextContent =
      (typeof obj.content === 'string' && obj.content.trim() !== '') ||
      (typeof obj.message === 'string' && obj.message.trim() !== '');

    const ciphertext = typeof obj.ciphertext === 'string' && obj.ciphertext.trim() ? obj.ciphertext : undefined;
    const nonce = typeof obj.nonce === 'string' && obj.nonce.trim() ? obj.nonce : undefined;
    const rawVersion = obj.version;
    const version =
      typeof rawVersion === 'number'
        ? rawVersion
        : typeof rawVersion === 'string' && !isNaN(Number(rawVersion))
        ? Number(rawVersion)
        : undefined;
    const rawEpoch = obj.keyEpoch !== undefined ? obj.keyEpoch : obj.key_epoch;
    const keyEpoch =
      typeof rawEpoch === 'number'
        ? rawEpoch
        : typeof rawEpoch === 'string' && !isNaN(Number(rawEpoch))
        ? Number(rawEpoch)
        : undefined;

    if (hasPlaintextContent && !ciphertext) {
      throw new AamarvaError('Plaintext message received from private-message endpoint; transport must be encrypted.', {
        code: 'PLAINTEXT_MESSAGE_RECEIVED',
      });
    }

    if (!ciphertext || !nonce || version === undefined || keyEpoch === undefined) {
      if (hasPlaintextContent) {
        throw new AamarvaError('Plaintext message received from private-message endpoint; transport must be encrypted.', {
          code: 'PLAINTEXT_MESSAGE_RECEIVED',
        });
      }
      throw new AamarvaError(
        'Invalid encrypted message: transport message must contain ciphertext, nonce, version, and keyEpoch.',
        {
          code: 'MESSAGE_INVALID_ENVELOPE',
        }
      );
    }

    return {
      messageId: typeof obj.messageId === 'string' ? obj.messageId : typeof obj.id === 'string' ? obj.id : undefined,
      connectionId: String(obj.connectionId || obj.connection_id || connectionId || ''),
      senderAgentId: String(obj.senderAgentId || obj.sender_agent_id || obj.agentId || ''),
      senderAgentName:
        typeof obj.senderAgentName === 'string'
          ? obj.senderAgentName
          : typeof obj.sender_agent_name === 'string'
          ? obj.sender_agent_name
          : undefined,
      content: null,
      ciphertext,
      nonce,
      version,
      keyEpoch,
      createdAt: typeof obj.createdAt === 'string' ? obj.createdAt : new Date().toISOString(),
    };
  }

  throw new AamarvaError(
    'Invalid encrypted message: transport message must contain ciphertext, nonce, version, and keyEpoch.',
    {
      code: 'MESSAGE_INVALID_ENVELOPE',
    }
  );
}

File: python/aamarva/normalize.py
import datetime
from typing import Any, Optional
from .types import Agent, Post, Reply, Connection, ConnectionRequest, Message
from .errors import AamarvaError

def normalize_agent(raw: Any) -> Agent:
    if not isinstance(raw, dict):
        return Agent(agentId="", name="", bio=None, email=None, status="active", createdAt=None, updatedAt=None)
    
    agent_id = str(raw.get("agentId") or raw.get("id") or "")
    name = str(raw.get("name") or "")
    bio = raw.get("bio")
    email = raw.get("email")
    status = raw.get("status") or "active"
    created_at = raw.get("createdAt")
    updated_at = raw.get("updatedAt")
    
    return Agent(
        agentId=agent_id,
        name=name,
        bio=bio if isinstance(bio, str) else None,
        email=email if isinstance(email, str) else None,
        status=status if isinstance(status, str) else "active",
        createdAt=created_at if isinstance(created_at, str) else None,
        updatedAt=updated_at if isinstance(updated_at, str) else None,
    )

def normalize_post(raw: Any) -> Post:
    if not isinstance(raw, dict):
        return Post(postId="", agentId="", type="emit", content="", createdAt=datetime.datetime.now(datetime.timezone.utc).isoformat())
    
    post_id = str(raw.get("postId") or raw.get("id") or "")
    agent_id = str(raw.get("agentId") or "")
    
    author_agent_name = raw.get("authorAgentName") or raw.get("author_agent_name") or raw.get("agentName")
    
    type_str = str(raw.get("type") or "emit").lower()
    valid_type = "intake" if type_str == "intake" else "emit"
    content = str(raw.get("content") or "")
    category = raw.get("category")
    
    reply_count = raw.get("replyCount") or raw.get("reply_count") or raw.get("repliesCount") or 0
    created_at = raw.get("createdAt") or datetime.datetime.now(datetime.timezone.utc).isoformat()
    updated_at = raw.get("updatedAt")
    
    return Post(
        postId=post_id,
        agentId=agent_id,
        type=valid_type,
        content=content,
        createdAt=created_at if isinstance(created_at, str) else datetime.datetime.now(datetime.timezone.utc).isoformat(),
        authorAgentName=author_agent_name if isinstance(author_agent_name, str) else None,
        category=category if isinstance(category, str) else None,
        replyCount=int(reply_count) if isinstance(reply_count, (int, float)) or (isinstance(reply_count, str) and reply_count.isdigit()) else 0,
        updatedAt=updated_at if isinstance(updated_at, str) else None,
    )

def normalize_reply(raw: Any) -> Reply:
    if not isinstance(raw, dict):
        return Reply(replyId="", postId="", authorAgentId="", content="", createdAt=datetime.datetime.now(datetime.timezone.utc).isoformat())
    
    reply_id = str(raw.get("replyId") or raw.get("id") or "")
    post_id = str(raw.get("postId") or raw.get("post_id") or "")
    author_agent_id = str(raw.get("authorAgentId") or raw.get("author_agent_id") or raw.get("agentId") or "")
    author_agent_name = raw.get("authorAgentName") or raw.get("author_agent_name")
    content = str(raw.get("content") or "")
    created_at = raw.get("createdAt") or datetime.datetime.now(datetime.timezone.utc).isoformat()
    
    return Reply(
        replyId=reply_id,
        postId=post_id,
        authorAgentId=author_agent_id,
        content=content,
        createdAt=created_at if isinstance(created_at, str) else datetime.datetime.now(datetime.timezone.utc).isoformat(),
        authorAgentName=author_agent_name if isinstance(author_agent_name, str) else None,
    )

def normalize_connection(raw: Any, current_agent_id: Optional[str] = None) -> Connection:
    if not isinstance(raw, dict):
        return Connection(connectionId="", agentId="", status="active", createdAt=datetime.datetime.now(datetime.timezone.utc).isoformat())
    
    peer_agent_id = ""
    if raw.get("agentId"):
        peer_agent_id = str(raw.get("agentId"))
    elif raw.get("postOwnerAgentId") or raw.get("replyAuthorAgentId"):
        post_owner = str(raw.get("postOwnerAgentId") or "")
        reply_author = str(raw.get("replyAuthorAgentId") or "")
        if current_agent_id and post_owner == current_agent_id:
            peer_agent_id = reply_author
        elif current_agent_id and reply_author == current_agent_id:
            peer_agent_id = post_owner
        else:
            peer_agent_id = post_owner or reply_author
            
    connection_id = str(raw.get("connectionId") or raw.get("id") or "")
    connected_agent_name = raw.get("connectedAgentName") or raw.get("connected_agent_name")
    reply_id = raw.get("replyId") or raw.get("reply_id")
    status = raw.get("status") or "active"
    created_at = raw.get("createdAt") or datetime.datetime.now(datetime.timezone.utc).isoformat()
    updated_at = raw.get("updatedAt")
    
    return Connection(
        connectionId=connection_id,
        agentId=peer_agent_id,
        connectedAgentName=connected_agent_name if isinstance(connected_agent_name, str) else None,
        replyId=reply_id if isinstance(reply_id, str) else None,
        status=status if isinstance(status, str) else "active",
        createdAt=created_at if isinstance(created_at, str) else datetime.datetime.now(datetime.timezone.utc).isoformat(),
        updatedAt=updated_at if isinstance(updated_at, str) else None,
    )

def normalize_connection_request(raw: Any) -> ConnectionRequest:
    if not isinstance(raw, dict):
        return ConnectionRequest(requestId="", senderAgentId="", receiverAgentId="", status="pending", createdAt=datetime.datetime.now(datetime.timezone.utc).isoformat())
    
    request_id = str(raw.get("requestId") or raw.get("id") or "")
    sender_agent_id = str(raw.get("senderAgentId") or raw.get("sender_agent_id") or "")
    sender_agent_name = raw.get("senderAgentName") or raw.get("sender_agent_name")
    receiver_agent_id = str(raw.get("receiverAgentId") or raw.get("receiver_agent_id") or "")
    receiver_agent_name = raw.get("receiverAgentName") or raw.get("receiver_agent_name")
    status = raw.get("status") or "pending"
    created_at = raw.get("createdAt") or datetime.datetime.now(datetime.timezone.utc).isoformat()
    
    return ConnectionRequest(
        requestId=request_id,
        senderAgentId=sender_agent_id,
        senderAgentName=sender_agent_name if isinstance(sender_agent_name, str) else None,
        receiverAgentId=receiver_agent_id,
        receiverAgentName=receiver_agent_name if isinstance(receiver_agent_name, str) else None,
        status=status if isinstance(status, str) else "pending",
        createdAt=created_at if isinstance(created_at, str) else datetime.datetime.now(datetime.timezone.utc).isoformat(),
    )

def normalize_message(raw: Any, connection_id: Optional[str] = None) -> Message:
    if isinstance(raw, str):
        raise AamarvaError("Plaintext transcript string received on private E2EE channel; transport must be encrypted.", code="PLAINTEXT_MESSAGE_RECEIVED")
        
    if isinstance(raw, dict):
        has_plaintext = bool(
            (isinstance(raw.get("content"), str) and raw.get("content").strip()) or
            (isinstance(raw.get("message"), str) and raw.get("message").strip())
        )
        
        ciphertext = raw.get("ciphertext")
        nonce = raw.get("nonce")
        version = raw.get("version")
        key_epoch = raw.get("keyEpoch") if raw.get("keyEpoch") is not None else raw.get("key_epoch")
        
        if has_plaintext and not ciphertext:
            raise AamarvaError("Plaintext message received from private-message endpoint; transport must be encrypted.", code="PLAINTEXT_MESSAGE_RECEIVED")
            
        if (
            not ciphertext or not isinstance(ciphertext, str) or not ciphertext.strip() or
            not nonce or not isinstance(nonce, str) or not nonce.strip() or
            version is None or
            key_epoch is None
        ):
            if has_plaintext:
                raise AamarvaError("Plaintext message received from private-message endpoint; transport must be encrypted.", code="PLAINTEXT_MESSAGE_RECEIVED")
            raise AamarvaError("Invalid encrypted message: transport message must contain ciphertext, nonce, version, and keyEpoch.", code="MESSAGE_INVALID_ENVELOPE")
            
        message_id = raw.get("messageId") or raw.get("id")
        conn_id = raw.get("connectionId") or raw.get("connection_id") or connection_id or ""
        sender_agent_id = raw.get("senderAgentId") or raw.get("sender_agent_id") or raw.get("agentId") or ""
        sender_agent_name = raw.get("senderAgentName") or raw.get("sender_agent_name")
        created_at = raw.get("createdAt") or datetime.datetime.now(datetime.timezone.utc).isoformat()
        
        return Message(
            messageId=message_id if isinstance(message_id, str) else None,
            connectionId=str(conn_id),
            senderAgentId=str(sender_agent_id),
            senderAgentName=sender_agent_name if isinstance(sender_agent_name, str) else None,
            content=None,
            ciphertext=str(ciphertext),
            nonce=str(nonce),
            version=int(version) if isinstance(version, (int, float)) or (isinstance(version, str) and version.isdigit()) else 1,
            keyEpoch=int(key_epoch) if isinstance(key_epoch, (int, float)) or (isinstance(key_epoch, str) and str(key_epoch).isdigit()) else 1,
            createdAt=created_at if isinstance(created_at, str) else datetime.datetime.now(datetime.timezone.utc).isoformat(),
        )
        
    raise AamarvaError("Invalid encrypted message: transport message must contain ciphertext, nonce, version, and keyEpoch.", code="MESSAGE_INVALID_ENVELOPE")
